model/riscv_vmem_utils.sail

```diff
@@ -0,0 +1,352 @@
+  var finished : bool = false;
+
+  var write_success : bool = true;
+  let vaddr = bits_of(vaddr);
+  repeat {
+    let offset = i;
+    let vaddr = vaddr + (offset * bytes);
+    match translateAddr(Virtaddr(vaddr), Write(Data)) {
+      TR_Failure(e, _) => return Err(Memory_Exception(Virtaddr(vaddr), e)),
+
+      /* NOTE: Currently, we only announce the effective address if address translation is successful.
+         This may need revisiting, particularly in the misaligned case. */
+      TR_Address(paddr, _) => {
+        /* If res is true, the load should be aligned, and this loop should only execute once */
+        if res & not(match_reservation(bits_of(paddr))) then {
+          write_success = false
+        } else match mem_write_ea(paddr, bytes, aq, rl, res) {
+          Err(e) => return Err(Memory_Exception(Virtaddr(vaddr), e)),
+
+          Ok(()) => {
+            let write_value = data[(8 * (offset + 1) * bytes) - 1 .. 8 * offset * bytes];
+            match mem_write_value(paddr, bytes, write_value, aq, rl, res) {
+              Err(e) => return Err(Memory_Exception(Virtaddr(vaddr), e)),
+              Ok(s)  => write_success = write_success & s,
+            }
+          }
+        }
+      }
+    };
+
+    if offset == last then {
+      finished = true
+    } else {
+      i = offset + step
+    }
+  } until finished;
+
+  Ok(write_success)
+}
+
+/****    External API    ****/
+
+function check_misaligned(vaddr : virtaddr, width : word_width) -> bool = {
+  if plat_enable_misaligned_access() then {
+    false
+  } else {
+    not(is_aligned_addr(vaddr, size_bytes(width)))
+  }
+}
+
+val vmem_read : forall 'width, 'width in {1, 2, 4, 8} & ('width <= xlen_bytes | 'width <= flen_bytes).
+  (regidx, xlenbits, int('width), bool, bool, bool) -> result(bits(8 * 'width), Retire_Failure)
+
+function vmem_read(rs, offset, width, aq, rl, res) = {
+  /* Get the address, X(rs1) + offset.
+   * Some extensions perform additional checks on address validity.
+   */
+  let vaddr : virtaddr = match ext_data_get_addr(rs, offset, Read(Data), width) {
+    Ext_DataAddr_OK(vaddr) => vaddr,
+    Ext_DataAddr_Error(e)  => return Err(Ext_DataAddr_Check_Failure(e)),
+  };
+
+  // TODO: Rationalize to use a single alignment definition if possible.
+  if res then {
+    /* "LR faults like a normal load, even though it's in the AMO major opcode space."
+     * - Andrew Waterman, isa-dev, 10 Jul 2018.
+     */
+    if   not(is_aligned_addr(vaddr, width))
+    then return Err(Memory_Exception(vaddr, E_Load_Addr_Align()));
+  } else {
+    if check_misaligned(vaddr, size_bytes(width))
+    then return Err(Memory_Exception(vaddr, E_Load_Addr_Align()));
+  };
+
+  /* If the load is misaligned or an allowed misaligned access, split into `n`
+     (single-copy-atomic) memory operations, each of `bytes` width. If the load is
+     aligned, then `n` = 1 and bytes will remain unchanged. */
+  let ('n, bytes) = split_misaligned(vaddr, width);
```

Comment:
We can work around the Rocq problem by changing `bytes` to `'bytes` - this gives `data` a printable type that can be used in rewrites.  (Rocq also needs a couple of termination measures, but I'll give those separately.)